commit 37ac3a18843c001f72523e455eeca128f4a89b86
Author: Antonio Terceiro <terceiro@softwarelivre.org>
Date:   Wed Mar 6 06:18:02 2013 +0800

    Backport multiarch support fixes from Ruby trunk

--- a/configure.in
+++ b/configure.in
@@ -231,6 +231,7 @@
 AC_DEFINE_UNQUOTED(RUBY_VERSION_NAME, RUBY_BASE_NAME"-"RUBY_LIB_VERSION !<verconf>!)
 
 AC_CANONICAL_TARGET
+test x"$target_alias" = x &&
 target_os=`echo $target_os | sed 's/linux-gnu$/linux/;s/linux-gnu/linux-/'`
 ac_install_sh='' # unusable for extension libraries.
 
@@ -2644,6 +2645,18 @@
     DLDLIBS="$DLDLIBS -lc"
     ])
 
+AC_ARG_ENABLE(multiarch,
+	      AS_HELP_STRING([--enable-multiarch], [enable multiarch compatible directories]),
+	      [multiarch=], [unset multiarch])
+if test ${multiarch+set}; then
+   AC_DEFINE(ENABLE_MULTIARCH)
+fi
+
+archlibdir='${libdir}/${arch}'
+sitearchlibdir='${libdir}/${sitearch}'
+archincludedir='${includedir}/${arch}'
+sitearchincludedir='${includedir}/${sitearch}'
+
 AC_ARG_WITH(soname,
 	AS_HELP_STRING([--with-soname=SONAME], [base name of shared library]),
 	[RUBY_SO_NAME=$withval], [RUBY_SO_NAME='$(RUBY_BASE_NAME)'])
@@ -2657,7 +2670,7 @@
 AC_ARG_ENABLE(shared,
        AS_HELP_STRING([--enable-shared], [build a shared library for Ruby]),
        [enable_shared=$enableval])
-libprefix='$(libdir)'
+libprefix=${multiarch+'$(archlibdir)'}${multiarch-'$(libdir)'}
 LIBRUBY_RELATIVE=${load_relative-no}
 AS_CASE("$enable_shared", [yes], [
   LIBRUBY='$(LIBRUBY_SO)'
@@ -2676,6 +2689,7 @@
     AS_CASE(["$libdir"], ['${exec_prefix}/'*], [libdir_basename=`basename "$libdir"`])
   fi
   AC_DEFINE_UNQUOTED(LIBDIR_BASENAME, ["${libdir_basename}"])
+  libdir_basename="${libdir_basename}"${multiarch+'/${arch}'}
 
   AS_CASE(["$target_os"],
     [freebsd*|dragonfly*], [],
@@ -2696,7 +2710,8 @@
 	LIBRUBY_DLDFLAGS='-Wl,-soname,lib$(RUBY_SO_NAME).so.$(MAJOR).$(MINOR)'" $LDFLAGS_OPTDIR"
 	LIBRUBY_ALIASES='lib$(RUBY_SO_NAME).so.$(MAJOR).$(MINOR) lib$(RUBY_SO_NAME).so'
 	if test "$load_relative" = yes; then
-	    LIBRUBY_RPATHFLAGS="'-Wl,-rpath,\$\${ORIGIN}/../${libdir_basename}'"
+	    libprefix="'\$\${ORIGIN}/../${libdir_basename}'"
+	    LIBRUBY_RPATHFLAGS="-Wl,-rpath,${libprefix}"
 	    LIBRUBY_RELATIVE=yes
 	fi
 	],
@@ -2810,7 +2825,7 @@
 ])
 if test "$enable_rpath" = yes; then
     test -z "$LIBRUBY_RPATHFLAGS" || LIBRUBY_RPATHFLAGS="$LIBRUBY_RPATHFLAGS "
-    LIBRUBY_RPATHFLAGS="$LIBRUBY_RPATHFLAGS${linker_flag}-R ${linker_flag}${libprefix} -L\$(libdir)"
+    LIBRUBY_RPATHFLAGS="$LIBRUBY_RPATHFLAGS${linker_flag}-R ${linker_flag}${libprefix} -L${libprefix}"
     LIBRUBYARG_SHARED="$LIBRUBY_RPATHFLAGS $LIBRUBYARG_SHARED"
     LIBRUBYARG_STATIC="$LIBRUBY_RPATHFLAGS $LIBRUBYARG_STATIC"
 fi
@@ -3205,15 +3220,6 @@
     rubyw_install_name='$(RUBYW_INSTALL_NAME)'
     ])
 
-AC_ARG_ENABLE(multiarch,
-	      AS_HELP_STRING([--enable-multiarch], [enable multiarch compatible directories]),
-	      [multiarch=], [unset multiarch])
-
-archlibdir='${libdir}/${arch}'
-sitearchlibdir='${libdir}/${sitearch}'
-archincludedir='${includedir}/${arch}'
-sitearchincludedir='${includedir}/${sitearch}'
-
 shvar_to_cpp() {
     var="$1" val="$2"
     exec_prefix_pat="`echo \"${exec_prefix}\" | sed 's/\\./\\\\./g'`"
@@ -3397,6 +3403,7 @@
     AC_DEFINE_UNQUOTED(RUBY_VENDOR_ARCH_LIB_FOR(arch), ${RUBY_VENDOR_ARCH_LIB_FOR} !<verconf>!)
 fi
 
+AC_SUBST(libdirname, ${multiarch+arch}libdir)
 AC_SUBST(archlibdir)dnl
 AC_SUBST(sitearchlibdir)dnl
 AC_SUBST(archincludedir)dnl
--- a/template/ruby.pc.in
+++ b/template/ruby.pc.in
@@ -15,15 +15,6 @@
 RUBY_SO_NAME=@RUBY_SO_NAME@
 RUBY_INSTALL_NAME=@RUBY_INSTALL_NAME@
 DEFFILE=@DEFFILE@
-LIBPATH=@LIBPATH@
-LIBRUBY_A=@LIBRUBY_A@
-LIBRUBY_SO=@LIBRUBY_SO@
-LIBRUBY=@LIBRUBY@
-LIBRUBYARG_SHARED=@LIBRUBYARG_SHARED@
-LIBRUBYARG_STATIC=@LIBRUBYARG_STATIC@
-LIBRUBYARG=@LIBRUBYARG@
-LIBS=@LIBS@
-DLDFLAGS=@DLDFLAGS@
 archlibdir=@archlibdir@
 sitearchlibdir=@sitearchlibdir@
 archincludedir=@archincludedir@
@@ -43,6 +34,15 @@
 rubyhdrdir=@rubyhdrdir@
 vendorhdrdir=@vendorhdrdir@
 sitehdrdir=@sitehdrdir@
+LIBPATH=@LIBPATH@
+LIBRUBY_A=@LIBRUBY_A@
+LIBRUBY_SO=@LIBRUBY_SO@
+LIBRUBY=@LIBRUBY@
+LIBRUBYARG_SHARED=@LIBRUBYARG_SHARED@
+LIBRUBYARG_STATIC=@LIBRUBYARG_STATIC@
+LIBRUBYARG=@LIBRUBYARG@
+LIBS=@LIBS@
+DLDFLAGS=@DLDFLAGS@
 
 Name: Ruby
 Description: Object Oriented Script Language
--- a/tool/rbinstall.rb
+++ b/tool/rbinstall.rb
@@ -304,7 +304,7 @@
 goruby_install_name = "go" + ruby_install_name
 
 bindir = CONFIG["bindir", true]
-libdir = CONFIG["libdir", true]
+libdir = CONFIG[CONFIG.fetch("libdirname", "libdir"), true]
 rubyhdrdir = CONFIG["rubyhdrdir", true]
 archhdrdir = CONFIG["rubyarchhdrdir"] || (rubyhdrdir + "/" + CONFIG['arch'])
 rubylibdir = CONFIG["rubylibdir", true]
